src/Preheat.cpp

```diff
@@ -0,0 +1,160 @@
+        config.time_to_heatup_1_degree[1] = 1.0 / (machine_nozzle_heat_up_speed - material_extrusion_cool_down_speed);
+        config.standby_temp = extruder_train.getSettingInSeconds("material_standby_temperature");
+
+        config.min_time_window = extruder_train.getSettingInSeconds("machine_min_cool_heat_time_window");
+
+        config.material_print_temperature = extruder_train.getSettingInDegreeCelsius("material_print_temperature");
+        config.material_print_temperature_layer_0 = extruder_train.getSettingInDegreeCelsius("material_print_temperature_layer_0");
+        config.material_initial_print_temperature = extruder_train.getSettingInDegreeCelsius("material_initial_print_temperature");
+        config.material_final_print_temperature = extruder_train.getSettingInDegreeCelsius("material_final_print_temperature");
+
+        config.flow_dependent_temperature = extruder_train.getSettingBoolean("material_flow_dependent_temperature"); 
+
+        config.flow_temp_graph = extruder_train.getSettingAsFlowTempGraph("material_flow_temp_graph"); // [[0.1,180],[20,230]]
+    }
+}
+
+double Preheat::getTimeToGoFromTempToTemp(int extruder, double temp_before, double temp_after, bool during_printing)
+{
+    Config& config = config_per_extruder[extruder];
+    double time;
+    if (temp_after > temp_before)
+    {
+        time = (temp_after - temp_before) * config.time_to_heatup_1_degree[during_printing];
+    }
+    else
+    {
+        time = (temp_before - temp_after) * config.time_to_cooldown_1_degree[during_printing];
+    }
+    return std::max(0.0, time);
+}
+
+double Preheat::getTemp(unsigned int extruder, double flow, bool is_initial_layer)
+{
+    if (is_initial_layer)
+    {
+        return config_per_extruder[extruder].material_print_temperature_layer_0;
+    }
+    return config_per_extruder[extruder].flow_temp_graph.getTemp(flow, config_per_extruder[extruder].material_print_temperature, config_per_extruder[extruder].flow_dependent_temperature);
+}
+
+Preheat::WarmUpResult Preheat::timeBeforeEndToInsertPreheatCommand_coolDownWarmUp(double time_window, unsigned int extruder, double temp0, double temp1, double temp2, bool during_printing)
+{
+    WarmUpResult result;
+    const Config& config = config_per_extruder[extruder];
+    double time_to_cooldown_1_degree = config.time_to_cooldown_1_degree[during_printing];
+    double time_to_heatup_1_degree = config.time_to_heatup_1_degree[during_printing];
+    result.total_time_window = time_window;
+
+    //                  ,temp2
+    //                 /                                    .
+    //     ,temp0     /                                     .
+    //      \ ' ' ' '/ ' ' '> outer_temp                    .
+    //       \______/                                       .
+    //               "-> temp1
+    //      ^^^^^^^^^^
+    //      limited_time_window
+    double outer_temp;
+    double limited_time_window;
+    if (temp0 < temp2)
```

Comment:
Shouldn't this be comparing `temp0` with `temp1`?

I guess that's precisely the problem with naming a variable `temp1`...